### ISSUE TITLE
HttpClient & ASP.NET Core instrumentation invokes Propagator.Inject/Extract when it should not

### ISSUE THREAD

**Issue by alanwest** [open-telemetry/opentelemetry-dotnet#1242]
I just spotted this in the HTTP instrumentation

https://github.com/open-telemetry/opentelemetry-dotnet/blob/dd4bd3e0bd6daa0ef114c34b1d216aa0e313f584/src/OpenTelemetry.Instrumentation.Http/Implementation/HttpHandlerDiagnosticListener.cs#L113-L116

This seems to suggest that we do not call `Inject` when the version of HttpClient being used is one that already propagates W3C trace context (.NET Core 3.0+ I believe). However, recent work was done defaulting to using the `CompositePropagator` (which propagates both trace context and baggage). So, by default, this instrumentation always invokes `Inject` now which is desirable for propagating baggage but redundant for propagating W3C trace context.

https://github.com/open-telemetry/opentelemetry-dotnet/blob/dd4bd3e0bd6daa0ef114c34b1d216aa0e313f584/src/OpenTelemetry.Instrumentation.Http/HttpClientInstrumentationOptions.cs#L39-L43

@eddynaka @CodeBlanch I imagine one of you may have a bit more context here since you both recently worked on propagating baggage in #1048.

**Comment by eddynaka** [comment]
In the "past" we had a method called `IsInjected` and we would only inject in `CompositePropagator` if we didn't have injected yet. But, that was removed. so I think that is a know "issue". Don't say it's a bug, because it would inject something that it already has.

**Comment by CodeBlanch** [comment]
@alanwest Your write-up is correct. We used to let .NET handle it, but when we added support for Baggage we switched to the CompositePropagator so it will now be firing by default where it didn't before.

It looks like .NET has an existence check:
https://github.com/dotnet/runtime/blob/632cdcab09fda363e460296a77be0be32096f512/src/libraries/System.Net.Http/src/System/Net/Http/DiagnosticsHandler.cs#L290

I'm not sure which fires first, us or them. Are you seeing duplicates? That would for sure be a bug. I think it's just a perf issue right now.

We could tinker with it and make the default just Baggage on .NET Core 3+? Basically to let .NET handle W3C traceparent+tracestate propagation and we can handle the extra stuff?

PS: I removed IsInjected because it was basically impossible to determine for something like Baggage which isn't necessarily present but the context overall could have still been injected. To say that another way, IsInjected made sense for TraceContext but not so much for CompositePropagator scenarios. It isn't part of the propagation spec, probably for that reason.

**Comment by alanwest** [comment]
> Are you seeing duplicates?

Nope. I just saw the code and thought it looked fishy, so figured I'd ask.

> I think it's just a perf issue right now.

Yes, seems like a possible minor perf issue if the `DiagnosticsHandler` does it work first, then the work we do is redundant.

> We could tinker with it and make the default just Baggage on .NET Core 3+?

This seems like a decent thing to explore.

At a minimum, I think we should do away with the `if (!(this.httpClientSupportsW3C && this.options.Propagator is TextMapPropagator))` check no longer really serves a purpose (and is a little confusing if you're not aware of the history 😄).

**Comment by alanwest** [comment]
@eddynaka @CodeBlanch 

Similar thing in ASP.NET Core instrumentation. Though I think the effect is a little worse in this case because an additional activity is created.

https://github.com/open-telemetry/opentelemetry-dotnet/blob/5e41c6232760a34d4b705aa5a625b7aa00d7ff2a/src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInListener.cs#L81-L103

**Comment by CodeBlanch** [comment]
@alanwest FYI it should only create that additional Activity if the extracted ActivityContext is different than the running one. That _should_ only happen if someone uses something like B3 which ASP.NET Core won't understand. Also, if you look in the Extract implementations, I added a short-circuit that quits out if the ActivityContext is already valid:

https://github.com/open-telemetry/opentelemetry-dotnet/blob/768eaa3c968197cd06283ef3d24a36d687ce3a84/src/OpenTelemetry.Api/Context/Propagation/TextMapPropagator.cs#L48-L52

Basically that was to reduce the perf hit on these double-extractions.

I'm not saying we can't improve it, just want to add all the relevant info 😄

**Comment by cijothomas** [comment]
Looks like we are doing the right things and there is no bug. Perhaps some room for minor perf optimizations. Removing bug tag.

**Comment by alanwest** [comment]
> Looks like we are doing the right things and there is no bug. Perhaps some room for minor perf optimizations. Removing bug tag.

Agreed, not a bug based on conversation. I opened an issue with the spec a while back (https://github.com/open-telemetry/opentelemetry-specification/issues/954) which may help facilitate any optimizations we might want to make.
